### DISCUSSION TITLE
type-hinting custom markers

### DISCUSSION THREAD

**Discussion by paxcodes** [pytest-dev/pytest#8293] [Q&A] [answered]
Is it correct that custom markers are expected to be type-hinted by the user (me) who is using pytest?

Given code like this, 

```python
@mark.real_http_call
class Test_Some_Service:
  pass
```

I am getting an error from Pyright on the `@mark...` line. 

> Untyped class decorator obscures type of class; ignoring decorator (reportUntypedClassDecorator)

Or is it already handled/typed by pytest somehow but pyright is not recognizing it like [in this issue](https://github.com/microsoft/pyright/issues/1434)?

**Comment by bluetech** [comment]
The situation is a bit complicated. pytest provides real proper typing for its own built-in marks, but cannot provide them for custom marks, at the moment. Nonetheless, there is still typing for the generic case, specifically (showing only the pertinent parts)

```py
Markable = TypeVar("Markable", bound=Union[Callable[..., object], type])

@attr.s(init=False, auto_attribs=True)
class MarkDecorator:
    # Type ignored because the overloads overlap with an incompatible
    # return type. Not much we can do about that. Thankfully mypy picks
    # the first match so it works out even if we break the rules.
    @overload
    def __call__(self, arg: Markable) -> Markable:  # type: ignore[misc]
        pass

    @overload
    def __call__(self, *args: object, **kwargs: object) -> "MarkDecorator":
        pass
```

This works for mypy, but maybe pyright doesn't like some of these tricks (quite understandable...).

**Comment by paxcodes** [comment, marked as answer]
Once I updated `pytest` to 6.2.2 the warnings were gone.
